src/orion/storage/legacy.py

```diff
@@ -135,10 +170,132 @@ def retrieve_result(self, trial, results_file=None, **kwargs):
+
+        update = dict(
+            status=status,
+            heartbeat=heartbeat,
+            experiment=trial.experiment
+        )
+        if trial.status == 'new':
+            update["start_time"] = datetime.datetime.utcnow()
+
+        elif status == 'completed':
+            update["end_time"] = datetime.datetime.utcnow()
+
+        rc = self._update_trial(trial, **update, where={'status': trial.status, '_id': trial.id})
+        trial.status = status
+
+        if not rc:
+            raise FailedUpdate()
+
+    def fetch_pending_trials(self, experiment):
+        """See :func:`~orion.storage.BaseStorageProtocol.fetch_pending_trials`"""
+        query = dict(
+            experiment=experiment._id,
+            status={'$in': ['new', 'suspended', 'interrupted']}
+        )
+        return self._fetch_trials(query)
+
+    def reserve_trial(self, experiment):
+        """See :func:`~orion.storage.BaseStorageProtocol.reserve_trial`"""
+        query = dict(
+            experiment=experiment._id,
+            status={'$in': ['interrupted', 'new', 'suspended']}
+        )
+        # read and write works on a single document
+        now = datetime.datetime.utcnow()
+        trial = self._db.read_and_write(
+            'trials',
+            query=query,
+            data=dict(
+                status='reserved',
+                start_time=now,
+                heartbeat=now
+            ))
+
+        if trial is None:
+            return None
+
+        return Trial(**trial)
+
+    def fetch_noncompleted_trials(self, experiment):
+        """See :func:`~orion.storage.BaseStorageProtocol.fetch_noncompleted_trials`"""
+        query = dict(
+            experiment=experiment._id,
+            status={'$ne': 'completed'}
+        )
+        return self._fetch_trials(query)
+
+    def fetch_completed_trials(self, experiment):
+        """See :func:`~orion.storage.BaseStorageProtocol.fetch_completed_trials`"""
+        query = dict(
+            experiment=experiment._id,
+            status='completed'
+        )
+        selection = {
+            'end_time': 1,
+            'results': 1,
+            'experiment': 1,
+            'params': 1
+        }
+        return self._fetch_trials(query, selection)
+
+    def count_completed_trials(self, experiment):
+        """See :func:`~orion.storage.BaseStorageProtocol.count_completed_trials`"""
+        query = dict(
+            experiment=experiment._id,
+            status='completed'
+        )
+        return self._db.count('trials', query)
+
+    def count_broken_trials(self, experiment):
+        """See :func:`~orion.storage.BaseStorageProtocol.count_broken_trials`"""
+        query = dict(
+            experiment=experiment._id,
+            status='broken'
+        )
+        return self._db.count('trials', query)
+
+    def update_heartbeat(self, trial):
+        """Update trial's heartbeat"""
+        return self._update_trial(trial, heartbeat=datetime.datetime.utcnow())
```

Comment:
This should verify first that `trial.status == 'reserved'`.

Reply:
track change the status to running. the heartbeat monitor check if the status is in a broken state instead

Reply:
The problem is `update_trial checks` that trial.status is still the same in the DB, but we could make a mistake and do `update_hearbeat` on a trial that is not reserved/running and that should not be possible. This is the legacy protocol, so we should never get running status from Track, only reserved.

Reply:
How was this one resolved?

Reply:
Note for reference: `update_heartbeat` now forces `update_trial(..., status='reserved')`.